### ISSUE TITLE
Bug: [no-unsafe-enum-comparison] false positive for ambient const enums

### ISSUE THREAD

**Issue by nhardy** [typescript-eslint/typescript-eslint#7908]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.2.2&showAST=types&fileType=.tsx&code=CYUwxgNghgTiAEYD2A7AzgF3iFBXAtvAKJ6EDeAUPPALJE0BCRASvALzwBE%2BI%2BARiBicKAXwoVQkWAmTosANygRcIAFzFSAbnEBLAGbwAFIuUI25rj36DOASniVRFIA&eslintrc=N4KABGBECmAeAu0B2ATAzpAXGA2uCUADgDYCuA5gJZKYAC8AnodGgMYBOlh8AtC8dXgB6dtFYB7ALaTkKaCkgAafBEgkK1Oo2ZtO3PmgFJhoidNnyeogI6lKnJOR7boPVgAsxAa2rlI%2BAF1lVUIAQ3Y0aHYAeW5KcSQMbFACInZxACsxeCwweHZSaHwAX2CoAuIWXJSCSHomFg4uXn5BISRxHlJE0IAzV2RSSTcpMM40BNyYdnT2fwhikGKgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkyJsqAhnugCYCysbArhmZTyYe6ADTgakEkOTQ8AOT6oAwgAt00ANYCwQkRIC%2BIA0A&tokens=false

### Repro Code

```typescript
declare const enum Enum {
  MEMBER = "member"
}

declare const value: Enum;

if (value === "member") {

}
```

### ESLint Config

```javascript
{
  "extends": [
    "plugin:@typescript-eslint/recommended",
    "plugin:@typescript-eslint/recommended-requiring-type-checking"
  ],
  "parserOptions": {
    "project": true
  },
  "rules": {
    "@typescript-eslint/no-unsafe-enum-comparison": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "isolatedModules": true,
    "strictNullChecks": true
  }
}
```

### Expected Result

There should be no lint error when comparing a value whose type is an ambient const enum with a literal value given that there is no enum that can be reference/imported to make this comparison.

### Actual Result

Lint error "The two values in this comparison do not have a shared enum type." was received. Suggested replacement with an enum comparison refers to a nonexistent enum (as the enum is an ambient const enum and is not a variable)

**Comment by bradzacher** [comment]
How are you consuming this type?
That value had to come from somewhere.
Or are you using an enum as a way of declaring a union type?

**Comment by nhardy** [comment]
I'm consuming the type from a library which declares a const enum which, by its nature is exported only as a type, not a value.

I noticed this issue specifically with the `ResourceType` enum in MapLibre. This is the type that I'm trying to compare with: https://github.com/maplibre/maplibre-gl-js/blob/6f4a2b06d60099949a9f0cf5bdf5a5f9e63e4b61/src/util/request_manager.ts#L6

**Comment by bradzacher** [comment]
Looking at their bundle - I can see the enum defined there:
https://unpkg.com/maplibre-gl@3.6.0/dist/maplibre-gl-dev.js
![image](https://github.com/typescript-eslint/typescript-eslint/assets/7462525/7473c141-f709-49bb-9202-17e959f11aad)
https://unpkg.com/maplibre-gl@3.6.0/dist/maplibre-gl.js
![image](https://github.com/typescript-eslint/typescript-eslint/assets/7462525/3290d3fe-1a37-47bf-a19b-503d52fb3be4)

It is actually an exported thing as part of the library.

Additionally - TS is actually smart enough to understand `const` enums from external declarations:
![image](https://github.com/typescript-eslint/typescript-eslint/assets/7462525/ebe24a15-f05d-4322-90a2-c056e3d22355)
It will use the declared values and inline them just like if they were defined inside your project as real const enums.

**Comment by nhardy** [comment]
Thanks for looking into this @bradzacher

> Looking at their bundle - I can see the enum defined there: https://unpkg.com/maplibre-gl@3.6.0/dist/maplibre-gl-dev.js
> 
> It is actually an exported thing as part of the library.

Ah I see. I believe that's because `maplibre-gl` is using `"isolatedModules": true` to produce their bundle, which means that const enums are in fact omitted (despite being listed as a const enum in the `.d.ts` files).

> Additionally - TS is actually smart enough to understand `const` enums from external declarations [...] It will use the declared values and inline them just like if they were defined inside your project as real const enums.

Thanks for this clarification. I was not aware about this behaviour of TypeScript. From what I can tell this is only true unless one of the following is set:
- `"isolatedModules": true`
- `"verbatimModuleSyntax": true`

In such cases, we're flagging that the code may be transpiled with a tool that is not type-aware like `tsc` is, and thus this automatic substitution of const enums may not take place and TypeScript produces the following error:

> Cannot access ambient const enums when 'isolatedModules' is enabled.ts(2748)

From this, I believe this bug should still be considered a false positive when the consuming application has `"isolatedModules": true` or `"verbatimModuleSyntax": true`, as a dependency may not include the enum at runtime (if the dependency itself was not using isolated modules or verbatim module syntax).

**Comment by bradzacher** [comment]
I'll leave this open as evaluating community engagement - but I my gut tells me this is just a super niche usecase that the vast, vast majority of user's would never encounter and thus is not worth the maintenance cost of implementation or the perf cost to check all the time.

I would, personally, be filing a bug with them upstream because realistically `const enum`s should never form part of a public API because they're supposed to be elided from the transpiled output.

**Comment by bradzacher** [comment]
But that's my point - nobody should be writing these by hand because you're creating a weird, pseudo-literal-union.

Such a construct would also provide [less safety than a literal union if you had any number members because it'd fall into the flag case.](https://www.typescriptlang.org/play?suppressImplicitAnyIndexErrors=true#code/KYDwDg9gTgLgBAE2AYwDYEMrDsiA7AZ3mDwFcBbOAOQrgG8504BeOARjgF8BYAKD9CRYOfEThlyALmq1WACgAMjAuIoAjYFACUAbj5A) I.e. It would allow any `number` typed value to be assigned.

And for a library - if you export a `const enum` then a `declare const enum` would slip into your `.d.ts` files - that's clearly a bug in the library because your users cannot use the construct you've defined - you get to fully use in the enum internally but externally users have to use raw literals - that's bad behaviour!

Which is why I say this is a case I think we should be fine not handling - because it shouldn't exist in good code and likely doesn't exist in the vast, vast majority of code.

**Comment by JoshuaKGoldberg** [comment]
👋 it's been over half a year since this was posted and we haven't seen any additional evidence of community desire for it. That leads us to believe that this is a pretty niche use case which we can't justify spending time creating/reviewing & then maintaining.

I'm going to go ahead and close this out as a wontfix. But if anybody feels strongly it should be reconsidered, please do post back in a new issue with a link to this one and an explanation of your use case. Cheers!
